java/ql/src/experimental/Security/CWE/CWE-600/UncaughtServletException.ql

```diff
@@ -0,0 +1,79 @@
+/**
+ * @name Uncaught Servlet Exception
+ * @description Uncaught exceptions in a servlet could leave a system in a vulnerable state, possibly resulting in denial-of-service attacks or the exposure of sensitive information disclosed in stack traces.
+ * @kind path-problem
+ * @id java/uncaught-servlet-exception
+ * @tags security
+ *       external/cwe-600
+ */
+
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.frameworks.Servlets
+import DataFlow::PathGraph
+
+/** The type `java.io.IOException`. */
+class IOException extends RefType {
+  IOException() { this.hasQualifiedName("java.io", "IOException") }
+}
+
+/** Check whether a given exception type is caught. */
+private predicate catchesEx(TryStmt t, RefType exType) {
+  exists(CatchClause cc, LocalVariableDeclExpr v |
+    t.getACatchClause() = cc and
+    cc.getVariable() = v and
+    v.getType().(RefType).getASubtype*() = exType //Detect the case that a subclass exception is thrown but its parent class is declared in the catch clause.
+  )
+}
+
+/** Servlet methods of `javax.servlet.http.HttpServlet`. */
+private predicate isServletMethod(Callable c) {
+  c.getDeclaringType() instanceof ServletClass and
+  c.getNumberOfParameters() = 2 and
+  c.getParameter(1).getType() instanceof HttpServletResponse and
```

Comment:
`ServletClass` matches `javax.servlet.Servlet` and `javax.servlet.http.HttpServlet`.
So to make `service` in line 43 cover `service(ServletRequest, ServletResponse)` as well, the following could be changed:
```suggestion
/** Servlet methods of `javax.servlet.Servlet` and subtypes. */
private predicate isServletMethod(Callable c) {
  c.getDeclaringType() instanceof ServletClass and
  c.getNumberOfParameters() = 2 and
  c.getParameter(1).getType() instanceof ServletResponse and
```

Though that would then also match other subtypes of `javax.servlet.Servlet`, which might or might not be desired.

Reply:
Thanks @Marcono1234 . I've updated the query to use the base type `ServletResponse`.